otumat/usage.py

```diff
@@ -0,0 +1,528 @@
+                                self.send()
+                            else:
+                                raise Exception('Unexpected server response...')
+                        except URLError as e:
+                            raise Exception('Connection refused when sending usage logs.')
+                        else:
+                            # insert successful, removing associated cached logs
+                            conn.execute('DELETE FROM event WHERE event_date < ?',
+                                         (current_time,))
+
+    def refresh_token(self):
+        """
+        Token refresh utility.
+        """
+        if self.config['collect']:
+            # build request to generate a new token
+            req = urllib_request.Request(
+                f"{self.config['host']}{self.config['refresh_route']}",
+                headers={'Authorization': f"""Basic {b64encode(f'''{self.config['client_id']}:{
+                    self.config['client_secret']}'''.encode('utf-8')).decode()}"""},
+                data=urlencode(
+                    dict(grant_type='refresh_token',
+                         refresh_token=self.config['refresh_token'])).encode('utf-8'))
+            try:
+                response = urllib_request.urlopen(req)
+            except HTTPError as e:
+                # access denied b/c refresh token has now expired
+                print('Usage upload connection has gone stale, requesting user to renew '
+                      'token manually...')
+                self.install()
+            except URLError as e:
+                raise Exception('Connection refused when requesting a new token.')
+            else:
+                # token returned successfully, update configuration
+                body = loads(response.read())
+                self.config['access_token'] = body['access_token']
+                self.config['expires_at'] = (datetime.utcnow().timestamp() +
+                                             int(body['expires_in']))
+                self.config['refresh_token'] = body['refresh_token']
+                self.config['scope'] = body['scope']
+                self.save_config()
+
+    def recurring_send(self, start: datetime, frequency='24h'):
+        """
+        Scheduler that uploads usage tracking data periodically.
+
+        :param start: Datetime to start upload schedule, if in past will ignore
+        :type start: datetime
+        :param frequency: Interval which to upload logs, defaults to '24h'
+        :type frequency: str, optional
+        """
+        # determine period in seconds
+        period, unit = [int(e) if e.isdigit() else e
+                        for e in findall(r'([0-9]+)([a-z]+)', frequency)[0]]
+        if unit == 'm':
+            period *= 60
+        elif unit == 'h':
+            period *= 60 * 60
+        elif unit == 'd':
+            period *= 24 * 60 * 60
```

Comment:
i would handle the `else` case here in case something other than `s` is used.

Reply:
Addressed